File: src/main/java/org/vstu/compprehension/controllers/interfaces/ExerciseController.java
package org.vstu.compprehension.controllers.interfaces;

import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;
import org.vstu.compprehension.dto.*;
import org.vstu.compprehension.dto.feedback.FeedbackDto;
import org.vstu.compprehension.dto.question.QuestionDto;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

public interface ExerciseController {

    @RequestMapping(value = {"/getExercises"}, method = { RequestMethod.GET })
    @ResponseBody
    List<Long> getExercises(HttpServletRequest request) throws Exception;

    /**
     * Get existing exercise attempt for current user
     * @param exerciseId Exercise id
     * @param request Current request
     * @return Existing exercise attempt or null
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/getExistingExerciseAttempt"}, method = { RequestMethod.GET })
    @ResponseBody
    ExerciseAttemptDto getExistingExerciseAttempt(Long exerciseId, HttpServletRequest request) throws Exception;

    /**
     * Create exercise attempt for current user
     * @param exerciseId Exercise id
     * @param request Current request
     * @return Created attempt
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/createExerciseAttempt"}, method = { RequestMethod.GET })
    @ResponseBody
    ExerciseAttemptDto createExerciseAttempt(Long exerciseId, HttpServletRequest request) throws Exception;

    /**
     * Returns statistics for all user's exercise attempt
     * @param exerciseId Exercise id
     * @return Statistics
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/getExerciseStatistics"}, method = { RequestMethod.GET })
    @ResponseBody
    ExerciseStatisticsItemDto[] getExerciseStatistics(Long exerciseId) throws Exception;

    /**
     * Add an answer to the question
     * @param interaction Interaction object
     * @param request Current request
     * @return Feedback
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/addQuestionAnswer"}, method = { RequestMethod.POST }, produces = "application/json",
            consumes = "application/json")
    @ResponseBody
    FeedbackDto addQuestionAnswer(@RequestBody InteractionDto interaction, HttpServletRequest request) throws Exception;

    /**
     * Generate new question for exercise attempt
     * @param exAttemptId Exercise attempt id
     * @param request Current request
     * @return Question
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/generateQuestion"}, method = { RequestMethod.GET })
    @ResponseBody
    QuestionDto generateQuestion(@RequestParam(name = "attemptId") Long exAttemptId, HttpServletRequest request) throws Exception;

    /**
     * Generate new supplementary question
     * @param questionRequest QuestionRequest
     * @param request Current request
     * @return Question
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/generateSupplementaryQuestion"}, method = { RequestMethod.POST })
    @ResponseBody
    QuestionDto generateSupplementaryQuestion(@RequestBody SupplementaryQuestionRequestDto questionRequest, HttpServletRequest request) throws Exception;

    /**
     * Get question by id
     * @param questionId Question Id
     * @param request Current request
     * @return Question
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/getQuestion"}, method = { RequestMethod.GET })
    @ResponseBody
    QuestionDto getQuestion(@RequestParam Long questionId, HttpServletRequest request) throws Exception;

    /**
     * Generate next correct answer
     * @param questionId Question Id
     * @param request Current request
     * @return Next correct answer
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/generateNextCorrectAnswer"}, method = { RequestMethod.GET })
    @ResponseBody
    FeedbackDto generateNextCorrectAnswer(@RequestParam Long questionId, HttpServletRequest request) throws Exception;

    /**
     * Root method for launching an exercise
     * @param exerciseId Exercise id
     * @param request Current request
     * @return Html page
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/pages/exercise"}, method = { RequestMethod.GET })
    String launch(Model model, Long exerciseId, HttpServletRequest request);

    /**
     * Load session info
     * @param request Current request
     * @return Session info
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/loadSessionInfo"}, method = { RequestMethod.GET })
    @ResponseBody
    SessionInfoDto loadSessionInfo(HttpServletRequest request) throws Exception;

    /**
     * Get current user info
     * @param request Current request
     * @return User info
     * @throws Exception Something got wrong
     */
    @RequestMapping(value = {"/getCurrentUser"}, method = { RequestMethod.GET })
    @ResponseBody
    UserInfoDto getCurrentUser(HttpServletRequest request) throws Exception;
}

File: src/main/js/client/controllers/exercise/exercise-controller.ts
import { injectable } from "tsyringe";
import { ExerciseAttempt, TExerciseAttempt, TOptionalExerciseAttemptResult } from "../../types/exercise-attempt";
import { ExerciseStatisticsItem, TExerciseStatisticsItems } from "../../types/exercise-statistics";
import { Feedback, TFeedback } from "../../types/feedback";
import { Interaction } from "../../types/interaction";
import { Question, TQuestion } from "../../types/question";
import { SessionInfo, TSessionInfo } from "../../types/session-info";
import { SupplementaryQuestionRequest } from "../../types/supplementary-question-request";
import { ajaxGet, ajaxPost, PromiseEither, RequestError } from "../../utils/ajax";


export interface IExerciseController {
    loadSessionInfo(): PromiseEither<RequestError, SessionInfo>;
    getQuestion(questionId: number): PromiseEither<RequestError, Question>;
    generateQuestion(attemptId: number): PromiseEither<RequestError, Question>;
    generateSupplementaryQuestion(questionRequest: SupplementaryQuestionRequest): PromiseEither<RequestError, Question>;
    generateNextCorrectAnswer(questionId: number): PromiseEither<RequestError, Feedback>;
    addQuestionAnswer(interaction: Interaction): PromiseEither<RequestError, Feedback> ;
    getExistingExerciseAttempt(exerciseId: number): PromiseEither<RequestError, ExerciseAttempt | null | undefined | ''>;
    createExerciseAttempt(exerciseId: number): PromiseEither<RequestError, ExerciseAttempt>;
    getExerciseStatistics(exerciseId: number): PromiseEither<RequestError, ExerciseStatisticsItem[]>;
}

@injectable()
export class ExerciseController implements IExerciseController {
    static endpointPath: string = ExerciseController.initEndpointPath();
    private static initEndpointPath() {
        const matches = /^\/(?:.+\/(?<!\/pages\/))?/.exec(window.location.pathname)
        if (matches) {
            return matches[0];
        }
        return "/";
    }

    loadSessionInfo(): PromiseEither<RequestError, SessionInfo> {
        return ajaxGet(`${ExerciseController.endpointPath}loadSessionInfo`, TSessionInfo);
    }

    getQuestion(questionId: number): PromiseEither<RequestError, Question> {
        return ajaxGet(`${ExerciseController.endpointPath}getQuestion?questionId=${questionId}`, TQuestion);
    }

    generateQuestion(attemptId: number): PromiseEither<RequestError, Question> {
        return ajaxGet(`${ExerciseController.endpointPath}generateQuestion?attemptId=${attemptId}`, TQuestion); 
    }

    generateSupplementaryQuestion(questionRequest: SupplementaryQuestionRequest): PromiseEither<RequestError, Question> {
        return ajaxPost(`${ExerciseController.endpointPath}generateSupplementaryQuestion`, questionRequest, TQuestion);
    }

    generateNextCorrectAnswer(questionId: number): PromiseEither<RequestError, Feedback> {
        return ajaxGet(`${ExerciseController.endpointPath}generateNextCorrectAnswer?questionId=${questionId}`, TFeedback);
    }

    addQuestionAnswer(interaction: Interaction): PromiseEither<RequestError, Feedback> {
        return ajaxPost(`${ExerciseController.endpointPath}addQuestionAnswer`, interaction, TFeedback);
    }

    getExistingExerciseAttempt(exerciseId: number): PromiseEither<RequestError, ExerciseAttempt | null | undefined | ''> {
        return ajaxGet(`${ExerciseController.endpointPath}getExistingExerciseAttempt?exerciseId=${exerciseId}`, TOptionalExerciseAttemptResult); 
    }

    createExerciseAttempt(exerciseId: number): PromiseEither<RequestError, ExerciseAttempt> {
        return ajaxGet(`${ExerciseController.endpointPath}createExerciseAttempt?exerciseId=${exerciseId}`, TExerciseAttempt); 
    }

    getExerciseStatistics(exerciseId: number): PromiseEither<RequestError, ExerciseStatisticsItem[]> {
        return ajaxGet(`${ExerciseController.endpointPath}getExerciseStatistics?exerciseId=${exerciseId}`, TExerciseStatisticsItems)
    }
}
